uast/original.go

```diff
@@ -0,0 +1,276 @@
+	if !ok {
+		return nil, ErrUnexpectedObject.New("map[string]interface{}", obj)
+	}
+
+	n := NewNode()
+	for k, o := range m {
+
+		switch ov := o.(type) {
+		case map[string]interface{}:
+			child, err := c.mapToNode(k, ov)
+			if err != nil {
+				return nil, err
+			}
+
+			n.Children = append(n.Children, child)
+		case []interface{}:
+			children, err := c.sliceToNodes(k, ov)
+			if err != nil {
+				return nil, err
+			}
+
+			n.Children = append(n.Children, children...)
+		default:
+			if err := c.addProperty(n, k, o); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	sort.Sort(byOffset(n.Children))
+	return n, nil
+}
+
+func (c *BaseOriginalToNoder) mapToNode(k string, obj map[string]interface{}) (*Node, error) {
+	n, err := c.toNode(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	n.Properties[InternalRoleKey] = k
+	return n, nil
+}
+
+func (c *BaseOriginalToNoder) sliceToNodes(k string, s []interface{}) ([]*Node, error) {
+	var ns []*Node
+	for _, v := range s {
+		n, err := c.toNode(v)
+		if err != nil {
+			return nil, err
+		}
+
+		n.Properties[InternalRoleKey] = k
+		ns = append(ns, n)
+	}
+
+	return ns, nil
+}
+
+func (c *BaseOriginalToNoder) addProperty(n *Node, k string, o interface{}) error {
+	switch {
+	case c.isTokenKey(k):
+		s, err := toString(o)
+		if err != nil {
+			return err
+		}
+
+		if n.Token != nil {
+			return fmt.Errorf("two token keys for same node: %s", k)
+		}
+
+		n.Token = &s
+	case c.InternalTypeKey == k:
+		s, err := toString(o)
+		if err != nil {
+			return err
+		}
+
+		if err := c.setInternalKey(n, s); err != nil {
+			return err
+		}
+
+		tk := c.syntheticToken(s)
+		if tk != nil {
+			if n.Token != nil {
+				return fmt.Errorf("two token keys for same node: %s", k)
+			}
+
+			n.Token = tk
+		}
+	case c.OffsetKey == k:
+		i, err := toUint32(o)
+		if err != nil {
+			return err
+		}
+
+		if n.StartPosition == nil {
+			n.StartPosition = &Position{}
+		}
+
+		n.StartPosition.Offset = &i
+	case c.LineKey == k:
+		i, err := toUint32(o)
+		if err != nil {
+			return err
+		}
+
+		if n.StartPosition == nil {
+			n.StartPosition = &Position{}
+		}
+
+		n.StartPosition.Line = &i
+	default:
+		s, err := toString(o)
+		if err != nil {
+			return err
+		}
+
+		n.Properties[k] = s
+	}
+
+	return nil
+}
+
+func (c *BaseOriginalToNoder) isTokenKey(key string) bool {
+	return c.TokenKeys != nil && c.TokenKeys[key]
+}
+
+func (c *BaseOriginalToNoder) syntheticToken(key string) *string {
+	if c.SyntheticTokens == nil {
+		return nil
+	}
+
+	t, ok := c.SyntheticTokens[key]
+	if !ok {
+		return nil
+	}
+
+	return &t
+}
+
+func (c *BaseOriginalToNoder) setInternalKey(n *Node, k string) error {
+	if n.InternalType != "" {
+		return fmt.Errorf("two internal keys for same node: %s, %s",
+			n.InternalType, k)
+	}
+
+	n.InternalType = k
+	return nil
+}
+
+func toString(v interface{}) (string, error) {
+	switch o := v.(type) {
+	case string:
+		return o, nil
+	case fmt.Stringer:
+		return o.String(), nil
+	case int:
```

Comment:
if this is for printing the things inside a tree, all numbers are `int64` after decoding json with a `map[string]interface{}`. Also, `bool` and `float64` are valid values.

You can directly stringify them all with `fmt.Sprint`